lib/migrations/accessible_objects_for_operations_function_cte.rb

```diff
@@ -0,0 +1,143 @@
+class AccessibleObjectsForOperationsFunctionCte < ActiveRecord::Migration[5.2]
+  def up
+    return unless PolicyMachineStorageAdapter.postgres?
+
+    execute <<~SQL.squish
+      CREATE OR REPLACE FUNCTION pm_accessible_objects_for_operations_cte(
+        user_id INT,
+        operations _TEXT,
+        field TEXT,
+        filters JSON DEFAULT '{}'
+      )
+      RETURNS TABLE (
+        unique_identifier varchar(255),
+        objects _varchar
+      ) AS $$
+      DECLARE
+        mergejoin TEXT;
+        filter_key TEXT;
+        filter_value TEXT;
+        filter_conditions TEXT = '';
+        user_attribute_table TEXT = 'user_attribute_ids';
+        query_text TEXT;
+      BEGIN
+        query_text := '
+        WITH RECURSIVE user_attribute_ids AS (
+          (
+            SELECT
+              child_id,
+              parent_id
+            FROM assignments
+            WHERE parent_id = $1
+          )
+          UNION ALL
+          (
+            SELECT
+              a.child_id,
+              a.parent_id
+            FROM assignments a
+            JOIN user_attribute_ids ua_id
+              ON ua_id.child_id = a.parent_id
+          )
+        ),';
+
+        IF filters IS NOT NULL AND filters::TEXT <> '{}' THEN
+          FOR filter_key, filter_value IN
+            SELECT * FROM json_each(filters)
+          LOOP
+            filter_conditions := filter_conditions || filter_key || ' = ' || filter_value || ' AND ';
+          END LOOP;
+
+          /* Chomp trailing AND */
+          filter_conditions := left(filter_conditions, -4);
+          /* Replace double quotes */
+          filter_conditions := replace(filter_conditions, '"', '''');
```

Comment:
this `''''` works? i'd think you'd need escaping

Reply:
Doubling single quotes is common in most RDBMS I've used. There are other options such as adding `E'\'foo\''`, character functions, formatting functions, etc

Reply:
roger just checking